core/src/main/java/com/linecorp/armeria/common/logging/RequestScopedMdc.java

```diff
@@ -319,35 +383,52 @@ public String get(String key) {
 
         @Override
         public Map<String, String> getCopyOfContextMap() {
-            final Map<String, String> threadLocalMap =
-                    firstNonNull(delegate.getCopyOfContextMap(), Collections.emptyMap());
+            final Map<String, String> threadLocalMap = getDelegateContextMap();
             final RequestContext ctx = RequestContext.currentOrNull();
             if (ctx == null) {
                 // No context available
-                return threadLocalMap;
+                if (threadLocalMap != null) {
+                    return maybeCloneThreadLocalMap(threadLocalMap);
+                } else {
+                    return Object2ObjectMaps.emptyMap();
+                }
             }
 
-            final Map<String, String> requestScopedMap =
-                    firstNonNull(getMap(ctx), Collections.emptyMap());
-            if (threadLocalMap.isEmpty()) {
+            // Retrieve the request-scoped properties.
+            // Note that this map is 1) unmodifiable and shared 2) or modifiable yet unshared,
+            // which means it's OK to return as it is or mutate it.
+            final Map<String, String> requestScopedMap = getAll(ctx);
+            if (threadLocalMap == null || threadLocalMap.isEmpty()) {
                 // No thread-local map available
                 return requestScopedMap;
             }
 
             // Thread-local map available
             if (requestScopedMap.isEmpty()) {
                 // Only thread-local map available
-                return threadLocalMap;
+                return maybeCloneThreadLocalMap(threadLocalMap);
             }
 
             // Both thread-local and request-scoped map available
-            final Map<String, String> merged =
-                    new Object2ObjectOpenHashMap<>(threadLocalMap.size() + requestScopedMap.size());
-            merged.putAll(threadLocalMap);
-            merged.putAll(requestScopedMap);
+            final Object2ObjectOpenHashMap<String, String> merged;
+            if (requestScopedMap instanceof Object2ObjectOpenHashMap) {
+                // Reuse the mutable copy returned by getAll() for less memory footprint.
+                merged = (Object2ObjectOpenHashMap<String, String>) requestScopedMap;
```

Comment:
Isn't this unmodifiable if the map is returned from here? https://github.com/line/armeria/pull/2729/files#diff-b35989b7cd6a82522044209c418a377dR193

Reply:
Yes. The map will not be an `Object2ObjectOpenHashMap` but be an `Object2ObjectMaps.Unmodifiable` or something. This code path is for the case where `getAll()` returns a mutable map, so we do not create two maps when merging twice.

Reply:
Ah I thought the condition was `requestScopedMap instanceof Object2ObjectMap`. Sorry about it. 😉